openghg/util/_metadata_util.py

```diff
@@ -0,0 +1,267 @@
+from typing import Any, Union, Iterable, List, Optional
+import logging
+import math
+
+
+__all__ = [
+    "null_metadata_values",
+    "not_set_metadata_values",
+    "remove_null_keys",
+    "check_number_match",
+    "check_str_match",
+    "check_value_match",
+    "check_not_set_value",
+    "get_overlap_keys",
+    "merge_dict",
+]
+
+
+logger = logging.getLogger("openghg.util")
+logger.setLevel(logging.INFO)  # Have to set level for logger as well as handler
+
+
+def null_metadata_values() -> List:
```

Comment:
I understand why this is a good suggestion but I feel like it may be an additional feature we maybe don't need. I'm a bit concerned about continually moving the code away from being clearly understood by the rest of the group so I am in favour of sometimes doing the *clear* thing in favour of the *right* thing. I'm not arguing very strongly for this but recognised syntax may be preferable in this case.

Reply:
I guess I don't see why it would be less clear. For instance np.nan is a constant. If a function only ever returns one value and that value isn't meant to be changed, then it might as well be an immutable value.

Reply:
It's more about known syntax - settings constants isn't something as common that people have used.

Reply:
We define type aliases and `TypeVar`s as module level variables/constants in several places in the code. Or in `util._user` we define `openghg_config_filename` as a module level constant. Defining a module level variable/constant is less esoteric than the code used to fetch parsers, which is treating modules like classes (which they are) and using `getattr`.

I'm not proposing something radical, it's basic syntax and functionality. Also, I'm not sure we should presuppose what other people in the group are capable of understanding. Sure, we should avoid unnecessarily complex code, but that's not the case here (assigning variables is surely less complex than defining functions).

Reply:
It's always a problem when writing code like to this to make sure we're not constantly moving goalposts and are keeping maintainability at the heart of it - its never really about what people are capable of understanding (I'm sure everyone is perfectly capable) more about the *number* of continuing new concepts being introduced. Some of this code being written in a vacuum has caused this problem before. I think in this case its probably not an issue but should always be a question we ask ourselves whenever we introduce something potentially *new* (and I use that term loosely), which is why I raised it.

For the code mentioned above this is usually around internal settings / type hints where it's unlikely to need to be changed and/or not directly affecting functionality. With this being around metadata, it felt more like we should be mindful of changes but we could easily argue this falls more into settings (when would we want to change this really) and could also argue it's not so much of a change.

In this case, is this a change you feel would strongly benefit the code? If so, great - let me know what you'd like it to change it to and I'll make the updates.

Reply:
I think the change I suggested would be an improvement: simply write `null_metadata_values = (None,)`, with the docstring from the function above that line. (Or write `tuple([None])` if you like, in this case, since there is only one value and the tuple syntax is weird in that case.) Same for the "not set" values function.

It's not crucial, and if you don't make the change, that's fine. I don't think the argument that it is a "new" concept really holds here, and seems like a bit of an overreaction to my suggestion. (Probably the reason it isn't so common in our code is that we keep moving things like this to configuration files, or that constants are being returned via functions when they don't really need to be, like the functions in `store/spec/_specification.py`.)

Reply:
I just wanted to clarify, I had thought the suggestion was the use of a special syntax (more complex than capital letters) and functionality around "constants" which would improve explicitness within the code but which we hadn't used elsewhere at this point. But I now believe I understand that this was a suggestion to use module-level variables instead of functions. In which case, you are correct that this is not a new concept and so this is something we would expect everyone to understand.

Thinking about it, I started considering the documentation aspect. Looking into docstrings specifically, though these can be written underneath variables and still interpreted by sphinx, in my opinion that makes it less clear and comments wouldn't be picked up. I do feel it's important that the purpose of these definitions are well understood so my preference would be in general to keep these things (i.e. global static data which would need an associated explanation) as functions with traditional docstrings, even if it would look simpler (ignoring the documentation part) with the single line. So I'm going to propose leaving it as it is for now.

Reply:
Okay, thanks for clarifying. I did mean a module level variable; I probably should have clarified that sooner.

I didn't realize that sphinx would pick up docstrings under variables. Still, keeping global static data in functions is a also a fine approach, and consistent with store.spec. Happy to accept your suggestion.